### PR TITLE
[Feature] Add the support for musa device support

### PR DESCRIPTION
**Pull request by hanhaowen-mt**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers. By the way, if you're not familiar with how to use pre-commit to fix lint issues or add unit tests, please refer to [Contributing to OpenMMLab](https://mmengine.readthedocs.io/en/latest/notes/contributing.html).

## Motivation

Please describe the motivation of this PR and the goal you want to achieve through this PR.

## Modification

Please briefly describe what modification is made in this PR.

## BC-breaking (Optional)

Does the modification introduce changes that break the backward-compatibility of the downstream repos?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
3. If the modification has potential influence on downstream projects, this PR should be tested with downstream projects, like MMDetection or MMPretrain.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

**Comment by zhouzaida** [comment]
HI @hanhaowen-mt , thanks for your contribution. Have you validated this support in downstream repos?

**Comment by hanhaowen-mt** [comment]
> HI @hanhaowen-mt , thanks for your contribution. Have you validated this support in downstream repos?

hi, I am still validating it yet

**Comment by hanhaowen-mt** [comment]
It is a draft, I am still struggling with many problems.....

**Comment by hanhaowen-mt** [comment]
@zhouzaida 请问一下为什么ci会显示torch无法被import？

**Comment by hanhaowen-mt** [comment]
目前mmcv已经适配完毕，但是还有些bug需要修复，由于mmengine是一个纯py代码库，所以我想可以先合入也不影响？

**Comment by hanhaowen-mt** [comment]
image
为什么会说import不了pytorch....
@zhouzaida

**Comment by zhouzaida** [comment]
https://github.com/open-mmlab/mmengine/blob/4a50213c6954d42536faa8b69b48bb6dd8f32f4f/.circleci/test.yml#L55 这一部分是用来测试没有 pytorch 的情况下，有些模块也可以正常使用

**Comment by hanhaowen-mt** [comment]
i/circleci: minimum_version_cpu 这个ci报错：
FAILED tests/test_hooks/test_empty_cache_hook.py::TestEmptyCacheHook::test_with_runner - AssertionError: 0 != 4
FAILED tests/test_runner/test_log_processor.py::TestLogProcessor::test_get_max_memory - AssertionError: Expected 'None' to have been called.
但是tests/test_hooks/test_empty_cache_hook.py我只加了一个注释，这里是不是误报错误？

File: mmengine/dist/dist.py [hunk 1]
```diff
@@ -776,8 +788,13 @@ def _gather_object(obj: Any,
...
     current_device = torch.device('cpu')
     is_nccl_backend = group_backend == torch_dist.Backend.NCCL
+    is_mccl_backend = group_backend == 'mccl'
     if is_nccl_backend:
-        current_device = torch.device('cuda', torch.cuda.current_device())
+        current_device = torch.device('', torch.cuda.current_device())
+        input_tensor = input_tensor.to(current_device)
+        local_size = local_size.to(current_device)
+    elif is_mccl_backend:
+        current_device = torch.device('', torch.musa.current_device())
```
Comment (zhouzaida) [inline mmengine/dist/dist.py:797]:
```suggestion
        current_device = torch.device('cuda', torch.cuda.current_device())
        input_tensor = input_tensor.to(current_device)
        local_size = local_size.to(current_device)
    elif is_mccl_backend:
        current_device = torch.device('musa', torch.musa.current_device())
```

File: mmengine/dist/utils.py [hunk 2]
```diff
@@ -12,6 +12,7 @@
 from torch import distributed as torch_dist
 from torch.distributed import ProcessGroup
 from mmengine.device import is_mlu_available, is_npu_available
+from mmengine.device import is_musa_available
```
Comment (zhouzaida) [inline mmengine/dist/utils.py:15]:
```suggestion
from mmengine.device import (is_mlu_available, is_npu_available, is_musa_available)
```

File: mmengine/hooks/empty_cache_hook.py [hunk 3]
```diff
@@ -50,6 +51,8 @@ def _after_iter(self,
         """
         if self._do_after_iter:
             torch.cuda.empty_cache()
+            if is_musa_available():
+                torch.musa.empty_cache()
```
Comment (zhouzaida) [inline mmengine/hooks/empty_cache_hook.py:55]: Whether the `torch.cuda.empty_cache` function can be called in musa?

File: mmengine/runner/log_processor.py [hunk 4]
```diff
@@ -231,6 +232,11 @@ def get_log_after_iter(self, runner, batch_idx: int,
             max_memory = self._get_max_memory(runner)
             log_str += f'memory: {max_memory}  '
             tag['memory'] = max_memory
+        # If musa is available, the max memory occupied should be calculated.
+        if is_musa_available():
```
Comment (zhouzaida) [inline mmengine/runner/log_processor.py:236]:
```suggestion
        elif is_musa_available():
```

File: mmengine/hooks/empty_cache_hook.py [hunk 5]
```diff
@@ -60,6 +63,8 @@ def _before_epoch(self, runner, mode: str = 'train') -> None:
         """
         if self._do_before_epoch:
             torch.cuda.empty_cache()
+            if is_musa_available():
+                torch.musa.empty_cache()
```
Comment (zhouzaida) [inline mmengine/hooks/empty_cache_hook.py:67]: same as above

File: mmengine/runner/utils.py [hunk 6]
```diff
@@ -70,6 +71,8 @@ def set_random_seed(seed: Optional[int] = None,
     torch.manual_seed(seed)
     # torch.cuda.manual_seed(seed)
     torch.cuda.manual_seed_all(seed)
+    if is_musa_available():
+        torch.musa.manual_seed_all(seed)
```
Comment (zhouzaida) [inline mmengine/runner/utils.py:75]: if is_cuda_available():
    torch.cuda.manual_seed_all(seed)
elif is_musa_available():
    torch.musa.manual_seed_all(seed)

File: mmengine/structures/base_data_element.py [hunk 7]
```diff
@@ -510,6 +510,17 @@ def cuda(self) -> 'BaseDataElement':
                 new_data.set_data(data)
         return new_data
 
+    # Tensor-like methods
+    def musa(self) -> 'BaseDataElement':
+        """Convert all tensors to GPU in data."""
```
Comment (zhouzaida) [inline mmengine/structures/base_data_element.py:515]:
```suggestion
        """Convert all tensors to musa in data."""
```

File: mmengine/utils/dl_utils/time_counter.py [hunk 8]
```diff
@@ -4,6 +4,7 @@
 
 import torch
 
+from mmengine.device import is_musa_available
```
Comment (zhouzaida) [inline mmengine/utils/dl_utils/time_counter.py:7]:
```suggestion
from mmengine.device import is_cuda_available, is_musa_available
```

File: mmengine/utils/dl_utils/time_counter.py [hunk 9]
```diff
@@ -84,15 +85,20 @@ def __call__(self, fn):
         def wrapper(*args, **kwargs):
             self.__count += 1
 
-            if self.with_sync and torch.cuda.is_available():
-                torch.cuda.synchronize()
+            if self.with_sync:
+                if torch.cuda.is_available():
```
Comment (zhouzaida) [inline mmengine/utils/dl_utils/time_counter.py:89]:
```suggestion
                if is_cuda_available():
```

File: mmengine/utils/dl_utils/time_counter.py [hunk 10]
```diff
@@ -84,15 +85,20 @@ def __call__(self, fn):
...
+                    torch.musa.synchronize()
             start_time = time.perf_counter()
 
             result = fn(*args, **kwargs)
 
-            if self.with_sync and torch.cuda.is_available():
-                torch.cuda.synchronize()
-
+            if self.with_sync:
+                if torch.cuda.is_available():
```
Comment (zhouzaida) [inline mmengine/utils/dl_utils/time_counter.py:98]:
```suggestion
                if is_cuda_available():
```

File: tests/test_runner/test_log_processor.py [hunk 11]
```diff
@@ -7,6 +7,7 @@
 import torch
 from parameterized import parameterized
 
+from mmengine.device import is_musa_available
```
Comment (zhouzaida) [inline tests/test_runner/test_log_processor.py:10]:
```suggestion
from mmengine.device import is_cuda_available, is_musa_available
```

File: tests/test_runner/test_log_processor.py [hunk 12]
```diff
@@ -113,7 +114,7 @@ def test_get_log_after_iter(self, by_epoch, mode, log_with_hierarchy):
                         f"time: {train_logs['time']:.4f}  "
                         f"data_time: {train_logs['data_time']:.4f}  ")
 
-            if torch.cuda.is_available():
+            if torch.cuda.is_available() or is_musa_available():
```
Comment (zhouzaida) [inline tests/test_runner/test_log_processor.py:117]:
```suggestion
            if is_cuda_available() or is_musa_available():
```

File: tests/test_runner/test_log_processor.py [hunk 13]
```diff
@@ -141,7 +142,7 @@ def test_get_log_after_iter(self, by_epoch, mode, log_with_hierarchy):
...
```
Comment (zhouzaida) [inline tests/test_runner/test_log_processor.py:145]:
```suggestion
            if is_cuda_available() or is_musa_available():
```

File: mmengine/utils/dl_utils/collect_env.py [hunk 14]
```diff
@@ -9,6 +10,7 @@
 import torch
 
 import mmengine
+from mmengine.device import is_musa_available
```
Comment (zhouzaida) [inline mmengine/utils/dl_utils/collect_env.py:13]:
```suggestion
from mmengine.device import is_cuda_available, is_musa_available
```

File: mmengine/utils/dl_utils/collect_env.py [hunk 15]
```diff
@@ -52,8 +58,9 @@ def collect_env():
     env_info['Python'] = sys.version.replace('\n', '')
 
     cuda_available = torch.cuda.is_available()
+    musa_available = is_musa_available()
     env_info['CUDA available'] = cuda_available
-
+    env_info['MUSA available'] = musa_available
     env_info['numpy_random_seed'] = np.random.get_state()[1][0]
 
     if cuda_available:
```
Comment (zhouzaida) [inline mmengine/utils/dl_utils/collect_env.py:66]:
```suggestion
    if is_cuda_available():
```

File: mmengine/utils/dl_utils/collect_env.py [hunk 16]
```diff
@@ -89,7 +96,23 @@ def collect_env():
                 except subprocess.SubprocessError:
                     nvcc = 'Not Available'
             env_info['NVCC'] = nvcc
+    elif musa_available:
```
Comment (zhouzaida) [inline mmengine/utils/dl_utils/collect_env.py:99]:
```suggestion
    elif is_cuda_available():
```

File: mmengine/utils/dl_utils/collect_env.py [hunk 17]
```diff
@@ -52,8 +58,9 @@ def collect_env():
     env_info['Python'] = sys.version.replace('\n', '')
 
     cuda_available = torch.cuda.is_available()
```
Comment (zhouzaida) [inline mmengine/utils/dl_utils/collect_env.py:60]:
```suggestion
    cuda_available = is_cuda_available()
```

File: mmengine/logging/logger.py [hunk 18]
```diff
@@ -398,22 +398,43 @@ def _get_device_id():
...
+                num_device = torch_musa.device_count()
+                musa_visible_devices = list(range(num_device))
+            else:
+                musa_visible_devices = musa_visible_devices.split(',')
+            try:
+                return int(musa_visible_devices[local_rank])
+            except ValueError:
+                # handle case for Multi-Instance GPUs
+                # see #1148 for details
+                return musa_visible_devices[local_rank]
```
Comment (zhouzaida) [inline mmengine/logging/logger.py:420]: Is it also necessary to try except for musa?

```python
return int(musa_visible_devices[local_rank])
```

**Comment by zhouzaida** [comment]
@hanhaowen-mt 你好，已完成 review，请根据 comment 进行修改。另外，想确认一下，此 PR 是否已在 mmdet 进行验证？

File: mmengine/runner/amp.py [hunk 19]
```diff
@@ -135,7 +135,13 @@ def autocast(device_type: Optional[str] = None,
...
         elif device_type == 'npu':
             pass
-
+        elif device_type == 'musa':
+            if dtype is None:
+                dtype = torch.get_autocast_gpu_dtype()
+            with torch.musa.amp.autocast(
+                    enabled=enabled, dtype=dtype, cache_enabled=cache_enabled):
+                yield
+                return
```
Comment (HAOCHENYE) [inline mmengine/runner/amp.py:144]: Is this `return` necessary

File: mmengine/runner/log_processor.py (hunk 4 above)
Comment (HAOCHENYE) [reply mmengine/runner/log_processor.py:236]: if is_cuda_available() or is_musa_available():

Comment (hanhaowen-mt) [reply mmengine/runner/log_processor.py:236]: done

File: mmengine/hooks/empty_cache_hook.py (hunk 5 above)
Comment (hanhaowen-mt) [reply mmengine/hooks/empty_cache_hook.py:67]: 这样写的话有一个和torch.cuda.empty_cache()的ut会挂掉，test_empty_cache_hook.py下的test_with_runner

File: mmengine/runner/utils.py (hunk 6 above)
Comment (hanhaowen-mt) [reply mmengine/runner/utils.py:75]: done

File: mmengine/logging/logger.py (hunk 18 above)
Comment (hanhaowen-mt) [reply mmengine/logging/logger.py:420]: done

File: mmengine/runner/amp.py (hunk 19 above)
Comment (hanhaowen-mt) [reply mmengine/runner/amp.py:144]: not necessary, already revised

File: mmengine/dist/dist.py [hunk 20]
```diff
@@ -776,8 +788,13 @@ def _gather_object(obj: Any,
     group_backend = get_backend(group)
     current_device = torch.device('cpu')
     is_nccl_backend = group_backend == torch_dist.Backend.NCCL
+    is_mccl_backend = group_backend == 'mccl'
     if is_nccl_backend:
-        current_device = torch.device('cuda', torch.cuda.current_device())
+        current_device = torch.device('', torch.cuda.current_device())
```
[the remark this answers is not in the document]
Comment (hanhaowen-mt) [reply mmengine/dist/dist.py]: 应该是误删了，已经修复

**Comment by hanhaowen-mt** [comment]
> @hanhaowen-mt 你好，已完成 review，请根据 comment 进行修改。另外，想确认一下，此 PR 是否已在 mmdet 进行验证？

修改完成，https://github.com/open-mmlab/mmdetection/pull/11352 mmdet的pr也提了，都是验证过的

File: mmengine/runner/amp.py (hunk 19 above)
Comment (zhouzaida) [reply mmengine/runner/amp.py]: > not necessary, already revised

@hanhaowen-mt this `return` should be kept.

File: mmengine/logging/logger.py [hunk 21]
```diff
@@ -398,22 +398,43 @@ def _get_device_id():
...
-            # handle case for Multi-Instance GPUs
-            # see #1148 for details
-            return cuda_visible_devices[local_rank]
+            import torch_musa
+            MUSA_AVAILABLE = True
+        except ImportError:
+            pass
+        if MUSA_AVAILABLE:
+            local_rank = int(os.getenv('LOCAL_RANK', '0'))
+            musa_visible_devices = os.getenv('MUSA_VISIBLE_DEVICES', None)
```
[the remark this answers is not in the document]
Comment (zhouzaida) [reply mmengine/logging/logger.py:409]: They are the same. `musa_visible_devices = os.getenv('MUSA_VISIBLE_DEVICES')` is equal to `musa_visible_devices = os.getenv('MUSA_VISIBLE_DEVICES', None)`

File: mmengine/runner/amp.py [hunk 22]
```diff
@@ -135,7 +135,12 @@ def autocast(device_type: Optional[str] = None,
 
         elif device_type == 'npu':
             pass
-
+        elif device_type == 'musa':
+            if dtype is None:
+                dtype = torch.get_autocast_gpu_dtype()
+            with torch.musa.amp.autocast(
+                    enabled=enabled, dtype=dtype, cache_enabled=cache_enabled):
+                yield
```
Comment (zhouzaida) [inline mmengine/runner/amp.py:143]: `return` should be added.

File: mmengine/runner/utils.py [hunk 23]
```diff
@@ -69,7 +70,10 @@ def set_random_seed(seed: Optional[int] = None,
     np.random.seed(seed)
     torch.manual_seed(seed)
     # torch.cuda.manual_seed(seed)
-    torch.cuda.manual_seed_all(seed)
+    if is_cuda_available():
+        torch.cuda.manual_seed_all(seed)
+    if is_musa_available():
```
Comment (zhouzaida) [inline mmengine/runner/utils.py:75]:
```suggestion
    elif is_musa_available():
```

File: mmengine/utils/dl_utils/collect_env.py [hunk 24]
```diff
@@ -24,6 +26,10 @@ def _get_cuda_home():
     return CUDA_HOME
 
 
+def _get_musa_home():
+    return os.environ.get('MUSA_HOME')
```
Comment (zhouzaida) [inline mmengine/utils/dl_utils/collect_env.py:30]:
```suggestion
    return os.getenv('MUSA_HOME')
```

File: mmengine/hooks/empty_cache_hook.py (hunk 3 above)
Comment (zhouzaida) [reply mmengine/hooks/empty_cache_hook.py:55]: 可以考虑改一下 `test_hooks/test_empty_cache_hook.py`. 只有 cuda is available 才测试 TestEmptyCacheHook

**Comment by zhouzaida** [comment]
@hanhaowen-mt 最后一轮 comment 了，可以改完再验证一下，没问题就可以合了

File: mmengine/runner/amp.py (hunk 19 above)
Comment (hanhaowen-mt) [reply mmengine/runner/amp.py]: yes, return should be kept....my bad...

**Comment by hanhaowen-mt** [comment]
> @hanhaowen-mt 最后一轮 comment 了，可以改完再验证一下，没问题就可以合了

搞定了，这个pr先合了吧

File: mmengine/hooks/empty_cache_hook.py (hunk 3 above)
Comment (zhouzaida) [reply mmengine/hooks/empty_cache_hook.py:55]: @hanhaowen-mt 麻烦这里也改一下

**Comment by zhouzaida** [comment]
@hanhaowen-mt 你好，[mmengine/hooks/empty_cache_hook.py](https://github.com/open-mmlab/mmengine/pull/1453/files/658a2f21da847e5cad113d2212f9754d271103ee#diff-96b33c7db70c9c88af8c7b144ce9f9be907bf645c36d608374102a46394f9167) 这里可以再改一下，然后单元测试根据有没有 cuda skip 掉就行可以。修完这个就可以合了。

File: mmengine/runner/utils.py (hunk 23 above)
Comment (zhouzaida) [inline mmengine/runner/utils.py:75]:
```suggestion
    if is_cuda_available():
        torch.cuda.manual_seed_all(seed)
    elif is_musa_available():
```

File: mmengine/hooks/empty_cache_hook.py [hunk 25]
```diff
@@ -69,4 +76,7 @@ def _after_epoch(self, runner, mode: str = 'train') -> None:
             mode (str): Current mode of runner. Defaults to 'train'.
         """
         if self._do_after_epoch:
-            torch.cuda.empty_cache()
+            if is_cuda_available():
+                torch.cuda.empty_cache()
+            if is_musa_available():
```
Comment (zhouzaida) [inline mmengine/hooks/empty_cache_hook.py:81]:
```suggestion
            elif is_musa_available():
```

File: mmengine/hooks/empty_cache_hook.py [hunk 26]
```diff
@@ -49,7 +50,10 @@ def _after_iter(self,
             mode (str): Current mode of runner. Defaults to 'train'.
         """
         if self._do_after_iter:
-            torch.cuda.empty_cache()
+            if is_cuda_available():
+                torch.cuda.empty_cache()
+            if is_musa_available():
```
Comment (zhouzaida) [inline mmengine/hooks/empty_cache_hook.py:55]:
```suggestion
            elif is_musa_available():
```

File: mmengine/hooks/empty_cache_hook.py [hunk 27]
```diff
@@ -59,7 +63,10 @@ def _before_epoch(self, runner, mode: str = 'train') -> None:
             mode (str): Current mode of runner. Defaults to 'train'.
         """
         if self._do_before_epoch:
-            torch.cuda.empty_cache()
+            if is_cuda_available():
+                torch.cuda.empty_cache()
+            if is_musa_available():
```
Comment (zhouzaida) [inline mmengine/hooks/empty_cache_hook.py:68]:
```suggestion
            elif is_musa_available():
```

**Review by zhouzaida — APPROVED** [review]

**Pull request merged**